### DISCUSSION TITLE
Delaying checking out a connection in sql_log_normalizer extension

### DISCUSSION THREAD

**Discussion by janko** [jeremyevans/sequel#1995] [General]
When using the [sequel-activerecord_connection](https://github.com/janko/sequel-activerecord_connection) extension in an Active Record app, I'm recommending people to use the `sql_log_normalizer` extension if they want Sequel's SQL logs to be normalized like Active Record's. 

I'm recommending initializing a Sequel database object in an initializer, and the `activerecord_connection` extension prevents the connection to be checked out (by setting `db.opts[:test]` to `false`), because the Rails apps needs to remain bootable without a database connection in order for rake tasks such as `db:create` and `db:drop` to work.

However, the `sql_log_normalizer` connection calls `Sequel::Database#literal` as soon as it's loaded, which checks out a connection and thus errors when the database doesn't exist. Would it be possible to somehow delay setting the SQL string escape type to when the connection would naturally be checked out?

**Comment by jeremyevans** [comment]
Then we would need to check for every logged SQL whether the `@sql_string_escape_type` has been set.  I'm sorry, but I'm not willing to compromise Sequel to make it work better with Rails.  My advice would be to fork the `sql_log_normalizer` extension, make the modifications you want, and ship that with `sequel-activerecord_connection`

**Reply by janko** [reply]
I was hoping we'd be able to utilize something like the `:after_connect` hook, in which case we'd only need to run this code when a new connection is opened. I viewed this change more like making the `sql_log_normalizer` extension support setting the `test: false` database option, which is not Rails-specific.

I'll see if there is something that can be done from within `sequel-activerecord_connection`, thanks for the input 🙂

**Reply by jeremyevans** [reply]
`:after_connect` is for per-connection changes, and this is a Database-level setting.

One possible option is not loading the `sql_log_normalizer` extension until you have a working connection.

**Reply by janko** [reply]
Interesting, I'm wondering how delaying loading the extension can be done from the user side. Is the below safe?

```rb
DB = Sequel.connect(..., test: false, after_connect: -> (conn) { DB.extension :sql_log_normalizer })
```

**Reply by jeremyevans** [reply]
I was assuming that there must be a point in Rails startup (maybe at the very end) where it would be safe to either assume a connection is loaded or force the connection to be loaded, at which point you could load the extension normally.  However, that may be a poor assumption.

I think your code would work, because attempt to load a Database extension multiple times is idempotent.

**Reply by janko** [reply]
Yeah, I think Rails just boots the entire application, even for rake tasks such as `db:create` and `db:drop`, which only require Active Record. I will investigate if there is a way to tell whether one of these two tasks are being executed, maybe that would be good enough. Thanks for the help!

**Comment by janko** [comment]
For rodauth-rails, I've decided to generate Sequel initialization directly inside the Rodauth configuration, which is autoloaded and thus isn't automatically loaded when the Rails app boots (unlike initializers). I set `keep_reference: false` to prevent database objects from accumulating when the code is reloaded in development.

```rb
class RodauthMain < Rodauth::Auth
  configure do
    # ...
    db Sequel.postgres(extensions: [:activerecord_connection, :sql_log_normalizer], keep_reference: false)
    # ...
  end
end
```

For other use cases where it would make sense to create a database object in an initializer, I will probably recommend something like:

```rb
DB = Sequel.postgres(extensions: :activerecord_connection)
begin
  DB.extension :sql_log_normalizer
rescue ActiveRecord::NoDatabaseError
end
```

The `:after_connect` solution I posted above is a bit convoluted just to accommodate for running `rails db:create` (`rails db:drop` already works).
